File: packages/shared-crypto/generate-keys.js
#!/usr/bin/env node

/**
 * Encryption Key Generator
 * =========================
 *
 * Generates secure encryption keys for the SAHOOL shared-crypto package.
 *
 * Usage:
 *   node generate-keys.js
 *   node generate-keys.js --format env
 *   node generate-keys.js --format json
 *
 * Author: SAHOOL Team
 */

const crypto = require('crypto');

// Parse command line arguments
const args = process.argv.slice(2);
const formatFlag = args.find(arg => arg.startsWith('--format'));
const format = formatFlag ? formatFlag.split('=')[1] : 'env';

// Generate keys
const encryptionKey = crypto.randomBytes(32).toString('hex');
const deterministicKey = crypto.randomBytes(32).toString('hex');
const hmacSecret = crypto.randomBytes(32).toString('hex');

// Output based on format
if (format === 'json') {
  // JSON format
  const keys = {
    ENCRYPTION_KEY: encryptionKey,
    DETERMINISTIC_ENCRYPTION_KEY: deterministicKey,
    HMAC_SECRET: hmacSecret,
    generated_at: new Date().toISOString(),
    warning: 'NEVER commit these keys to version control. Store in a secure secret manager.'
  };
  console.log(JSON.stringify(keys, null, 2));
} else {
  // ENV format (default)
  console.log('# ========================================');
  console.log('# SAHOOL Encryption Keys');
  console.log('# Generated:', new Date().toISOString());
  console.log('# ========================================');
  console.log('');
  console.log('# ⚠️  WARNING: Never commit these keys to version control!');
  console.log('# Store them securely in AWS Secrets Manager, Azure Key Vault, or similar.');
  console.log('');
  console.log('# Primary Encryption Key (for standard encryption)');
  console.log(`ENCRYPTION_KEY=${encryptionKey}`);
  console.log('');
  console.log('# Deterministic Encryption Key (for searchable fields)');
  console.log(`DETERMINISTIC_ENCRYPTION_KEY=${deterministicKey}`);
  console.log('');
  console.log('# HMAC Secret (for data integrity verification)');
  console.log(`HMAC_SECRET=${hmacSecret}`);
  console.log('');
  console.log('# ========================================');
  console.log('# Next Steps:');
  console.log('# 1. Copy these keys to your .env file');
  console.log('# 2. Store backups in a secure location');
  console.log('# 3. Never share keys in plain text');
  console.log('# 4. Set up key rotation schedule');
  console.log('# ========================================');
}

File: packages/shared-crypto/generate-keys.py
#!/usr/bin/env python3

"""
Encryption Key Generator (Python)
==================================

Generates secure encryption keys for the SAHOOL shared-crypto package.

Usage:
    python3 generate-keys.py
    python3 generate-keys.py --format env
    python3 generate-keys.py --format json

Author: SAHOOL Team
"""

import json
import os
import sys
from datetime import datetime


def generate_keys():
    """Generate encryption keys."""
    return {
        "ENCRYPTION_KEY": os.urandom(32).hex(),
        "DETERMINISTIC_ENCRYPTION_KEY": os.urandom(32).hex(),
        "HMAC_SECRET": os.urandom(32).hex(),
    }


def output_env_format(keys):
    """Output keys in .env format."""
    print("# ========================================")
    print("# SAHOOL Encryption Keys")
    print(f"# Generated: {datetime.now().isoformat()}")
    print("# ========================================")
    print("")
    print("# ⚠️  WARNING: Never commit these keys to version control!")
    print("# Store them securely in AWS Secrets Manager, Azure Key Vault, or similar.")
    print("")
    print("# Primary Encryption Key (for standard encryption)")
    print(f"ENCRYPTION_KEY={keys['ENCRYPTION_KEY']}")
    print("")
    print("# Deterministic Encryption Key (for searchable fields)")
    print(f"DETERMINISTIC_ENCRYPTION_KEY={keys['DETERMINISTIC_ENCRYPTION_KEY']}")
    print("")
    print("# HMAC Secret (for data integrity verification)")
    print(f"HMAC_SECRET={keys['HMAC_SECRET']}")
    print("")
    print("# ========================================")
    print("# Next Steps:")
    print("# 1. Copy these keys to your .env file")
    print("# 2. Store backups in a secure location")
    print("# 3. Never share keys in plain text")
    print("# 4. Set up key rotation schedule")
    print("# ========================================")


def output_json_format(keys):
    """Output keys in JSON format."""
    output = {
        **keys,
        "generated_at": datetime.now().isoformat(),
        "warning": "NEVER commit these keys to version control. Store in a secure secret manager.",
    }
    print(json.dumps(output, indent=2))


def main():
    """Main function."""
    # Parse command line arguments
    format_type = "env"
    if len(sys.argv) > 1:
        for arg in sys.argv[1:]:
            if arg.startswith("--format="):
                format_type = arg.split("=")[1]

    # Generate keys
    keys = generate_keys()

    # Output based on format
    if format_type == "json":
        output_json_format(keys)
    else:
        output_env_format(keys)


if __name__ == "__main__":
    main()
